Code/Framework/AzCore/AzCore/DOM/DomValue.cpp

```diff
@@ -0,0 +1,1241 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#pragma once
+
+#include <AzCore/DOM/DomValue.h>
+#include <AzCore/DOM/DomValueWriter.h>
+#include <AzCore/std/smart_ptr/make_shared.h>
+
+namespace AZ::Dom
+{
+    template<class T>
+    AZStd::shared_ptr<T>& CheckCopyOnWrite(AZStd::shared_ptr<T>& refCountedPointer)
+    {
+        if (refCountedPointer.use_count() > 1)
+        {
+            AZStd::shared_ptr<T> newPointer = AZStd::allocate_shared<T>(AZStdAlloc<ValueAllocator>());
+            *newPointer = *refCountedPointer;
+            refCountedPointer = AZStd::move(newPointer);
+        }
+        return refCountedPointer;
+    }
+
+    Node::Node(AZ::Name name)
+        : m_name(name)
+    {
+    }
+
+    AZ::Name Node::GetName() const
+    {
+        return m_name;
+    }
+
+    void Node::SetName(AZ::Name name)
+    {
+        m_name = name;
+    }
+
+    Object::ContainerType& Node::GetProperties()
+    {
+        return m_properties;
+    }
+
+    const Object::ContainerType& Node::GetProperties() const
+    {
+        return m_properties;
+    }
+
+    Array::ContainerType& Node::GetChildren()
+    {
+        return m_children;
+    }
+
+    const Array::ContainerType& Node::GetChildren() const
+    {
+        return m_children;
+    }
+
+    Value::Value(AZStd::shared_ptr<const AZStd::string> string)
+        : m_value(string)
+    {
+    }
+
+    Value::Value()
+    {
+    }
+
+    Value::Value(const Value& value)
+        : m_value(value.m_value)
+    {
+    }
+
+    Value::Value(Value&& value) noexcept
+    {
+        operator=(value);
+    }
```

Comment:
This is backwards, don't do this.

This will attempt to invoke the copy assignment operator of the `Value` class, and because this code would not work if for some reason `operator=` invokes a virtual function as the vtable for the `Value` class isn't available in the constructor.

This should instead use the constructor member initializer list.

```suggestion
    Value::Value(Value&& value) noexcept
        : m_value{ AZStd::move(value.m_value) }
    {
    }
```

Reply:
I swapped in a memcpy implementation that profiled slightly faster in the end, but kept it separate from operator= to hopefully keep things idiomatic